packages/block-editor/src/components/colors/use-colors.js

```diff
@@ -0,0 +1,156 @@
+import ContrastChecker from '../contrast-checker';
+import InspectorControls from '../inspector-controls';
+import { useBlockEditContext } from '../block-edit';
+
+const ColorPanel = ( {
+	title,
+	colorSettings,
+	colorPanelProps,
+	contrastCheckerProps,
+	components,
+	panelChildren,
+} ) => (
+	<PanelColorSettings
+		title={ title }
+		initialOpen={ false }
+		colorSettings={ colorSettings }
+		{ ...colorPanelProps }
+	>
+		{ contrastCheckerProps &&
+			components.map( ( Component ) => (
+				<ContrastChecker
+					key={ Component.displayName }
+					textColor={ Component.color }
+					{ ...contrastCheckerProps }
+				/>
+			) ) }
+		{ typeof panelChildren === 'function' ?
+			panelChildren( components ) :
+			panelChildren }
+	</PanelColorSettings>
+);
+const InspectorControlsColorPanel = ( props ) => (
+	<InspectorControls>
+		<ColorPanel { ...props } />
+	</InspectorControls>
+);
+
+export default function __experimentalUseColors(
+	colorConfigs,
+	{
+		panelTitle = __( 'Color Settings' ),
+		colorPanelProps,
+		contrastCheckerProps,
+		panelChildren,
+	} = {
+		panelTitle: __( 'Color Settings' ),
+	},
+	deps = []
+) {
+	const { attributes, setAttributes } = useBlockEditContext();
+
+	const createComponent = useMemo(
+		() =>
+			memoize(
+				( attribute, color ) => ( { children } ) =>
+					// Clone children, setting the style attribute from the color configuration,
+					// if not already set explicitly through props.
+					Children.map( children, ( child ) =>
+						cloneElement( child, {
+							className: color ?
+								`${ child.props.className } has-${ kebabCase( attribute ) }` :
+								child.props.className,
+							style: { [ attribute ]: color, ...child.props.style },
+						} )
+					),
+				{ maxSize: colorConfigs.length }
+			),
+		[ colorConfigs.length ]
+	);
+	const createSetColor = useMemo(
+		() =>
+			memoize( ( name ) => ( newColor ) => setAttributes( { [ name ]: newColor } ), {
+				maxSize: colorConfigs.length,
+			} ),
+		[ setAttributes, colorConfigs.length ]
+	);
+
+	return useMemo( () => {
+		const colorSettings = [];
+
+		const components = colorConfigs.reduce( ( acc, colorConfig ) => {
+			if ( typeof colorConfig === 'string' ) {
+				colorConfig = { name: colorConfig };
+			}
+			const {
+				name, // E.g. 'backgroundColor'.
+				attribute = name, // E.g. 'backgroundColor'.
+
+				panelLabel = startCase( name ), // E.g. 'Background Color'.
+				componentName = panelLabel.replace( /\s/g, '' ), // E.g. 'BackgroundColor'.
+
+				color = colorConfig.color,
+				colors,
+			} = {
+				...colorConfig,
+				color: attributes[ colorConfig.name ],
+			};
+
+			// We memoize the non-primitives to avoid unnecessary updates
+			// when they are used as props for other components.
+			acc[ componentName ] = createComponent( attribute, color );
+			acc[ componentName ].displayName = componentName;
+			acc[ componentName ].color = color;
+			acc[ componentName ].setColor = createSetColor( name );
```

Comment:
The combination of memoize with useMemo works and should be performant, but it took me some time to understand it.
Can we extract the setters and components creation to its own useMemo?
e.g: something like:
```
const colorSetters = useMemo(
	() => {
	return reduce( colorNames, ( accumulator, colorName ) => {
		accumulator[ colorName ] = ( newColor ) => setAttributes( { [ colorName ]: newColor } 			);
	} ),
	[ colorNames, setAttributes ]
);
const colorComponents = useMemo(...
```
Color colorNames would porbably come from another small useMemo hook.

We then would have a final useMemo that depends on colorSetters, colorComponents and probably colorConfig and generates the final object.
This approach probably has the disadvantage that as the colorComponents will depend on color value if the value of color changes all the colorComponents will be regenerated instead of just the color component of the color that changed, but I don't think this has a noticeable performance impact and the code simplification it brings may worth it.

Reply:
I don't see how that code would look simpler. If anything, it creates more code and a soup of hook dependencies which are already not the easiest to understand.